### ISSUE TITLE
var_acf untested, unclear

### ISSUE THREAD

**Issue by jbrockmendel** [statsmodels/statsmodels#4572]
AFAICT the only tests for [VARProcess.acf](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/tsa/vector_ar/var_model.py#L105) are smoketests.  Putting together a case and manually calculating the autocovariances I'm getting numbers off by 3+ orders of magnitude.  So the first thing to check is that I'm not making some dumb mistake by forgetting what "autocovariance" means.

The `_var_acf` docstring reads:
```
    """
    Compute autocovariance function ACF_y(h) for h=1,...,p
    Notes
    -----
    Lütkepohl (2005) p.29
    """
```

For those of us without a copy of Lutkepohl lying around, is there another reference available?

**Comment by josef-pkt** [comment]
VAR/VECM is based almost only on Lütkepohl's book.
There are a few other books that include VAR, but I never looked at any. (A long time ago I read some sections in Hamilton.)
I don't think it's in any articles unless they are very old, the Lütkepohl articles that I read were more on surveys of VAR/VECM basics or special topics for it.

It might be possible to cross-check against the impulse response function, i.e. ma representation, but I would need pencil and paper to figure that out.

(aside: I had a hard time writing unit tests for similar properties of an ARMA process a long time ago, because there were no reference results in other packages available at the time or that I knew of.)

**Comment by jbrockmendel** [comment]
I'd rather not post a link, but the second Google result for "Lutkepohl 2005" looks like it might be the relevant book.  The vech/unvech/duplication_matrix notation Lutkepohl uses is going to take some getting used to.

>>> It might be possible to cross-check against the impulse response function, i.e. ma representation, but I would need pencil and paper to figure that out.

I'll make a PR with a real test once I get this sorted out.  The low-tech approach is to simulate the process and calculate the sample autocovariances.

**Comment by josef-pkt** [comment]
> The low-tech approach is to simulate the process and calculate the sample autocovariances.

I do this kind of simulation checks very often, but adding those to the unit test will soon make the test time very long. I prefer regression test after having checked with some heavier simulation results to verify that it looks "reasonable". (and consistency results and law of large numbers are always useful for checking)

**Comment by ChadFulton** [comment]
I have a function for computing VAR autocovariances that you can check against (although also not unit tested):

`tsa.statespace.tools._compute_multivariate_acovf_from_coefficients`

It is unit tested in:

`tsa.statespace.tests.test_tools.test_multivariate_acovf`

And you could use those unit tests if you like.

**Comment by josef-pkt** [comment]
>  there’s a count_model test that takes something like 140 seconds locally

You could open a separate issue if there are problems with a count model, adding more details

the entire discrete/tests only take 130 second on my computer (python 3.4)
891 passed, 1 skipped in 129.13 seconds
and for test_count_models.py
82 passed in 62.30 seconds

**Comment by jbrockmendel** [comment]
The number I quoted was off.  `TestZeroInflatedNegativeBinomialP_predict2.setup_class` takes 69.6 seconds on my machine.  Total calls to `GenericZeroInflated.fit` take 147.1s, clocking in at 5.1 seconds apiece.

While this is something I'd like to improve, its not a big deal and can get its own issue if it becomes one.  The point was just that a hypothetical 0.1s `simulate_var` call is not something to worry about at this juncture.  We've got an important part of the code that is essentially untested; that's the issue here.

**Comment by jbrockmendel** [comment]
Simulation results using the following code, tried with `k_ar=2, 4`:

```
# proc is a VARProcess with neqs=3, k_ar of 2 or 4
acovs = []
for k in range(10**4):
    sim = proc.simulate_var(steps=10**4)
    ans = np.array((10, neq, neqs))
    for t in range(10):
        # manually calculate 
        acov = np.cov(sim[t:].T, sim[:len(sim)-t].T)
        ans[t] = acov[:neqs, neqs:].T
    
    acovs.append(ans)

acovs = np.array(acovs)
mean = acovs.mean(0)
std = acovs.std(0)
stderr = std / np.sqrt(len(acovs))

acf = proc.acf(9)
ts = (mean - acf) / stderr
```

Results for `ts` in the `k_ar=2` case:

```
>>> ts[0]
array([[-1.86892606, -1.76499752, -1.78309391],
       [-1.76499752, -1.95396964, -1.06690155],
       [-1.78309391, -1.06690155, -2.31149213]])
>>> ts[1]
array([[-1.21927345, -0.21646114, -1.75177846],
       [-0.95909816, -0.32264154, -1.66760468],
       [-1.26769696, -0.86342303, -1.4165468 ]])
>>> ts[2]
array([[ -617.47103207, -1043.64022168,  -536.26338671],
       [  715.5411308 ,   164.47843769,   401.3218848 ],
       [-1288.26083394,  -997.75301738, -1256.86390154]])
>>> ts[3]
array([[-464.21187481, -247.32454381, -800.37281852],
       [ -77.25154332,  145.00926529, -339.17386614],
       [-374.41897359, -346.8433262 , -610.94499575]])
```

In the `k_ar=4` case:
```
>>> ts[0]
array([[-3.89968625, -2.5312309 , -3.43955847],
       [-2.5312309 , -2.63076744, -1.63326932],
       [-3.43955847, -1.63326932, -3.42702767]])
>>> ts[1]
array([[-1.43952735, -2.61771183, -0.87142639],
       [-2.33696172, -2.88859626, -1.84364914],
       [-0.79342493, -2.25407772, -0.56797275]])
>>> ts[2]
array([[-1.30076371, -1.76867021, -1.16728744],
       [-1.37508468, -1.53802663, -1.47087028],
       [-1.14451351, -1.25148128, -1.21639465]])
>>> ts[3]
array([[-1.73470011, -1.66563617, -1.80305086],
       [-1.77552925, -1.44080897, -2.08768739],
       [-1.63848443, -1.92341812, -1.30454781]])
>>> ts[4]
array([[-477.40327482, -699.28397656, -405.44985537],
       [ 448.83192431,   99.50454984,  261.29543772],
       [-903.59791128, -674.64281183, -875.99192821]])
>>> ts[5]
array([[-410.39539453, -206.29422346, -528.29201178],
       [ -91.47008463,   79.57869435, -222.68422552],
       [-404.78848723, -293.85551652, -433.88031189]])
```

This suggests that up through `lag=k_ar` the results are about right, and things go off the rails after that.  Note that `acf[:k_ar]` is calculated using `var_model._var_acf`, while `acf[k_ar:]` is calculated using yule-walker equations.  Nothing immediately stands out to me in the yule-walker implementation, so this could use another pair of eyeballs.

**Comment by ChadFulton** [comment]
It looks to me like the issue is that `_var_acf` is returning the autocovariances 0, -1, -2, ..., -p rather than 0, 1, ..., p. (i.e. they're not getting transposed right after the `unvec`)

**Comment by jbrockmendel** [comment]
@ChadFulton to make sure I understand the suggestion:

Current:
```
    acf = tsatools.unvec(vecACF)
    acf = acf[:k].T.reshape((p, k, k))
```

Are you suggesting this should be:
```
    acf = tsatools.unvec(vecACF).T
    acf = acf[:k].T.reshape((p, k, k))
```
?

**Comment by jbrockmendel** [comment]
Based on a quick eye-balling, the following change in `VARProcess.acf` seems to fix the mismatch:

```
                result[h] += np.dot(coefs[j], result[h - j - 1])
```
becomes
```
                result[h] += np.dot(coefs[j], result[h - j - 1].T).T
```

**Comment by ChadFulton** [comment]
```python
result[h] += np.dot(coefs[j], result[h - j - 1].T).T
```

That will continue to compute the negative autocovariances, and `_var_acf` will continue to be incorrect (relative to what its docstring says, anyway). Instead:

Current:

```python
acf = tsatools.unvec(vecACF)
acf = acf[:k].T.reshape((p, k, k))
```

Desired:

```python
acf = tsatools.unvec(vecACF)
acf = np.array(np.split(acf[:k], p, axis=1))
```

**Comment by ChadFulton** [comment]
> That will continue to compute the negative autocovariances

To clarify, what I meant it will compute the autocovariances for `h=-1, -2, ...`. This may slightly confusing because `h` is defined here as in Lutkepohl, and he defines the autocovariance as:

E[(y_t - mu) (y_{t-h} - mu)']

So when h = 1, if we are following Lutkepohl, then we should be computing Cov(y_t, y_{t-1}). This is done by the formula I gave. I think that this is the more usual timing convention in the definition of the autocovariance function for VARs.

On the other hand, if you instead wanted to compute Cov(y_t, y_{t+h}) when h=1, then you would want to use the formula you wrote down.

**Comment by jbrockmendel** [comment]
Mostly good news.  Overnight I ran 10^5 simulations to manually calculate the autocovariance (with a `k_ar=2` case) uses Chad's suggested change.  The results:

- There is no longer a big jump in the mismatch at `k_ar=2`
- t-statistics comparing the analytic acf with the sample acf are all between -2.6 and -9.1
- but with the large sample, the difference are of order 1e-7 to 1e-9
- It tentatively looks like the difference is driven by sampling variation in `E[y_t]E[y_{t-h}]`.  When I calculate the base truth as `cov[h] = np.dot(y[h:].T, y[:-h]) / (len(y) - h) - mu**2` instead of `cov[h] = np.cov(...)`, the t-stats look OK.  This was with a smaller sample; I'm running the bigger sample now.

**Comment by jbrockmendel** [comment]
Shoot.  With the 10^5 sample we're back to having 12 (of 90) of the tstats outside the (-1.96, 1.96) range.  That could be swept under the rug with a "well maybe HAC adjustments would get that down to 5" were it not for the fact that all 9 of the `h=1` tstats are among those 12.

**Comment by ChadFulton** [comment]
I'm not sure what the conclusion is here, but I think this at least indicates a bug in `_var_acf` that should be fixed along the lines I suggested? Are you concerned that we still haven't got the correct formula?

**Comment by jbrockmendel** [comment]
Your suggested fix to `_var_acf` is definitely an improvement.  The remaining mismatch must be numerical error somewhere, but I'm not sure where.

**Comment by ChadFulton** [comment]
> The remaining mismatch must be numerical error somewhere, but I'm not sure where.

I was surprised that I wasn't able to find another package computing theoretical VAR autocovariances that we could check against. Do you know of one? Somewhere in Matlab maybe?

**Comment by bashtage** [comment]
Fix in #5501
